allow setting flags via env vars

linki opened the pull request:
~~Just an example how we would do it today. I suggest we use a different flags processing library.~~

~~It should support env var processing by default. Molecule uses env vars to configure their controller and we want ext dns to be as easy a replacement as it can be.~~

Updates flags to final `v0.3` so we can move to `beta`. Allows using env vars to configure all flags. Switched out pflag for kingpin in order to achieve that.

Includes patch updates to some unrelated packages (looking mad at you, `glide`) . Should be fine, though.

ideahitme commented on pkg/apis/externaldns/types.go:
```diff
@@ -51,6 +54,11 @@ func NewConfig() *Config {
 func (cfg *Config) ParseFlags(args []string) error {
+	defaultDryRun, err := strconv.ParseBool(os.Getenv("DRY_RUN"))
+	if err != nil {
+		defaultDryRun = true
```
it is probably safe in this case, but in general should probably fail if conversion fails. Or use another flag lib :D

linki replied:
I think I'll do that. Env var processing is not high prio atm. I'll leave this open till then. Molecule also requires ALIAS and multi-zone. So no need to rush this here.

linki commented:
@ideahitme I changed the flags to the new defaults for `v0.3`. Also switched out `pflag` for `kingpin` in order to better support env vars. There's no an env var for each flag. Let me know what you think.

Raffo commented on pkg/apis/externaldns/types.go:
```diff
@@ -61,28 +58,34 @@ func NewConfig() *Config {
+	app.Flag("policy", "Modify how DNS records are sychronized between sources and providers; possible choices: <sync|upsert-only> default: sync").Default("sync").StringVar(&cfg.Policy)
+	app.Flag("compatibility", "Process annotation semantics from legacy implementations; possible choices: <mate|molecule>, default: none").StringVar(&cfg.Compatibility)
+	app.Flag("metrics-address", "Specify were to serve the metrics and health check endpoint").Default(defaultMetricsAddress).StringVar(&cfg.MetricsAddress)
+	app.Flag("log-format", "The format in which log messages are printed").Default(defaultLogFormat).StringVar(&cfg.LogFormat)
```
Do you also want to say what are the possible formats? Like this I wouldn't know.

hjacobs replied:
Yes, please allow me as a user to have a complete picture how to use the CLI options by just using `--help`.

ideahitme commented on pkg/apis/externaldns/types.go:
```diff
@@ -61,28 +78,47 @@ func NewConfig() *Config {
-	flags.StringVar(&cfg.FqdnTemplate, "fqdn-template", "", `fallback template to generate DNS name if annotation is missing; 
-		e.g. --fqdn-template={{.Name}}-{{.Namespace}}.example.org will use service/ingress name/namespace`)
-	return flags.Parse(args)
+	app := kingpin.New("external-dns", "ExternalDNS synchronizes exposed Kubernetes Services and Ingresses with DNS providers.")
```
👍

ideahitme commented on pkg/apis/externaldns/types.go:
```diff
@@ -61,28 +78,47 @@ func NewConfig() *Config {
+	// Flags related to Kubernetes
+	app.Flag("master", "The Kubernetes API server to connect to; defaults to auto-detection from the environment").Default(defaultConfig.Master).StringVar(&cfg.Master)
+	app.Flag("kubeconfig", "Retrieve target cluster configuration from a Kubernetes configuration file; defaults to auto-detection from the environment").Default(defaultConfig.KubeConfig).StringVar(&cfg.KubeConfig)
+	app.Flag("namespace", "Limit sources of endpoints to a specific namespace; defaults to all").Default(defaultConfig.Namespace).StringVar(&cfg.Namespace)
```
shouldn't namespace flag belong to `processing sources` group below ?

linki replied:
Yes, thought about that too. I'll move it.

ideahitme commented:
lgtm would be good to see small note on changelog if something relevant has changed

ideahitme commented on pkg/apis/externaldns/types.go:
```diff
@@ -61,28 +78,47 @@ func NewConfig() *Config {
+	app.Flag("namespace", "Limit sources of endpoints to a specific namespace; defaults to all").Default(defaultConfig.Namespace).StringVar(&cfg.Namespace)
+
+	// Flags related to processing sources
+	app.Flag("source", "The resource types that are queried for endpoints").EnumsVar(&cfg.Sources, "service", "ingress")
```
how does this work ? `--source=service,ingress` or `--source=service -- source=ingress` ?

linki replied:
I'll document it. We could write  a custom parser upstream that can do both at the same time :)

linki commented:
@ideahitme I further sanitized the flags and updated the changelog

linki commented:
@Raffo @hjacobs flags are now much more descriptive:

```
usage: external-dns --source=source --provider=provider [<flags>]

ExternalDNS synchronizes exposed Kubernetes Services and Ingresses with DNS providers.

Flags:
  --help                     Show context-sensitive help (also try --help-long and --help-man).
  --version                  Show application version.
  --master=""                The Kubernetes API server to connect to (default: auto-detect)
  --kubeconfig=""            Retrieve target cluster configuration from a Kubernetes configuration file (default: auto-detect)
  --source=source ...        The resource types that are queried for endpoints; specify multiple times for multiple sources (required, options: service, ingress)
  --namespace=""             Limit sources of endpoints to a specific namespace (default: all namespaces)
  --fqdn-template=""         A templated string that's used to generate DNS names from sources that don't define a hostname themselves (optional)
  --compatibility=           Process annotation semantics from legacy implementations (optional, options: mate, molecule)
  --provider=provider        The DNS provider where the DNS records will be created (required, options: aws, google)
  --google-project=""        When using the Google provider, specify the Google project (required when --provider=google)
  --domain-filter=""         Limit possible target zones by a domain suffix (optional)
  --policy=sync              Modify how DNS records are sychronized between sources and providers (default: sync, options: sync, upsert-only)
  --registry=txt             The registry implementation to use to keep track of DNS record ownership (default: txt, options: txt, noop)
  --txt-owner-id="default"   When using the TXT registry, a name that identifies this instance of ExternalDNS (default: default)
  --txt-prefix=""            When using the TXT registry, a custom string that's prefixed to each ownership DNS record (optional)
  --interval=1m0s            The interval between two consecutive synchronizations in duration format (default: 1m)
  --once                     When enabled, exits the synchronization loop after the first iteration (default: disabled)
  --dry-run                  When enabled, prints DNS record changes rather than actually performing them (default: disabled)
  --log-format=text          The format in which log messages are printed (default: text, options: text, json)
  --metrics-address=":7979"  Specify were to serve the metrics and health check endpoint (default: :7979)
  --debug                    When enabled, increases the logging output for debugging purposes (default: disabled)
```
